### PR TITLE
Measure GIL contention in benchmarks

### PR DESCRIPTION
**Pull request by pentschev**
Add argument to measure and report GIL contention in benchmarks. This may provide some useful insight when optimizing asyncio-related changes.

Sample results observed are below.

<details><summary>Small messages (1 B)</summary>

```
$ python -m ucp.benchmarks.send_recv --backend ucp-core --n-bytes 1 --n-iter 100_000 --no-detailed-report --report-gil-contention
...
GIL contention            | 0.688429594039917

$ python -m ucp.benchmarks.send_recv --backend ucp-async --n-bytes 1 --n-iter 100_000 --no-detailed-report --report-gil-contention
...
GIL contention            | 0.4772491455078125

$ UCXPY_NON_BLOCKING_MODE=1 python -m ucp.benchmarks.send_recv --backend ucp-async --n-bytes 1 --n-iter 100_000 --no-detailed-report --report-gil-contention
...
GIL contention            | 0.8763511180877686
```

</details>

<details><summary>Medium messages (1 MiB)</summary>

```
$ python -m ucp.benchmarks.send_recv --backend ucp-core --n-bytes 1MiB --n-iter 100 --no-detailed-report --report-gil-contention
...
GIL contention            | 0.1721574366092682

$ python -m ucp.benchmarks.send_recv --backend ucp-async --n-bytes 1MiB --n-iter 100 --no-detailed-report --report-gil-contention
...
GIL contention            | 0.25215959548950195

$ UCXPY_NON_BLOCKING_MODE=1 python -m ucp.benchmarks.send_recv --backend ucp-async --n-bytes 1MiB --n-iter 100 --no-detailed-report --report-gil-contention
...
GIL contention            | 0.2693021595478058

```

</details>

<details><summary>Large messages (1 GiB)</summary>

```
$ python -m ucp.benchmarks.send_recv --backend ucp-core --n-bytes 1 GiB --n-iter 10 --no-detailed-report --report-gil-contention
...
GIL contention            | 0.5908554792404175

$ python -m ucp.benchmarks.send_recv --backend ucp-async --n-bytes 1GiB --n-iter 10 --no-detailed-report --report-gil-contention
...
GIL contention            | 0.6251853108406067

$ UCXPY_NON_BLOCKING_MODE=1 python -m ucp.benchmarks.send_recv --backend ucp-async --n-bytes 1GiB --n-iter 10 --no-detailed-report --report-gil-contention
...
GIL contention            | 0.832139790058136
```

</details>

### CODE REVIEW

**Comment by quasiben** [comment]
This is really cool!  Can you describe what the output results are  ?  Is it time, a ratio, etc ?

**Comment by pentschev** [comment]
It is a ratio of the perceived time the user code (in this case, UCX-Py send/recv messages) had the GIL. Thus 0.0 would mean the user code never took the GIL, whereas 1.0 would mean the user code had the GIL for the entirety of time without releasing it.

**Comment by jacobtomlinson** [comment]
This looks neat! The Dask dashboard has a plot for this too. There were some performance implications with `gilknocker==0.3.0` but [seems much better](https://github.com/dask/distributed/pull/7651#discussion_r1138286148) in `0.4.0`.

File: ucp/benchmarks/backends/tornado.py [hunk 1]
```diff
@@ -99,4 +107,11 @@ async def run(self) -> bool:
             stop = monotonic()
             if i >= self.args.n_warmup_iter:
                 times.append(stop - start)
+
+        if self.args.report_gil_contention:
+            knocker.stop()
+            knocker.stop()
```
Comment (wence-) [inline ucp/benchmarks/backends/tornado.py:113]:
```suggestion
```

I assume we only need to stop once.

File: ucp/benchmarks/backends/ucp_core.py [hunk 2]
```diff
@@ -292,10 +299,15 @@ def op_completed():
             if i >= self.args.n_warmup_iter:
                 times.append(stop - start)
 
+        if self.args.report_gil_contention:
+            knocker.stop()
+            knocker.stop()
```
Comment (wence-) [inline ucp/benchmarks/backends/ucp_core.py:304]:
```suggestion
```

File: ucp/benchmarks/send_recv.py [hunk 3]
```diff
@@ -337,6 +347,15 @@ def parse_args():
     if args.backend != "ucp-core" and args.delay_progress:
         raise RuntimeError("`--delay-progress` requires `--backend=ucp-core`")
 
+    if args.report_gil_contention:
+        try:
+            import gilknocker  # noqa: F401
+        except ImportError:
+            raise RuntimeError(
+                "Could not import `gilknocker`. Make sure it is installed or "
+                "remove the `--measure-gil-contention` argument."
```
Comment (wence-) [inline ucp/benchmarks/send_recv.py:356]:
```suggestion
                "remove the `--report-gil-contention` argument."
```

**Review by wence- — APPROVED** [review]
Minor nits, looks good though.

Interested to know how contended things are.

**Comment by pentschev** [comment]
> Interested to know how contended things are.

Not sure if you mean something different, but I [reported contention numbers in the description](https://github.com/rapidsai/ucx-py/pull/937#issue-1639304458). Did you overlook that or are you asking for something else?

**Comment by wence-** [comment]
> > Interested to know how contended things are.
> 
> Not sure if you mean something different, but I [reported contention numbers in the description](https://github.com/rapidsai/ucx-py/pull/937#issue-1639304458). Did you overlook that or are you asking for something else?

Oh no, I'm just blind...

**Comment by pentschev** [comment]
Thanks @wence- !

**Pull request merged**